### PR TITLE
Avoid polymorphic compare in Ident

### PR DESCRIPTION
**Pull request by lpw25**
The `Ident` module is quite important for the type-checker's performance. Currently it uses polymorphic compare/equality unnecessarily in two ways:
1. It uses it to compare strings rather than the type-specialised version
2. It uses it to implement `Ident.same` where it could instead just check the stamps of the identifiers

This PR removes these uses.

### CODE REVIEW

File: typing/ident.ml [hunk 1]
```diff
@@ -93,13 +93,16 @@ let equal i1 i2 =
...
 
-let same i1 i2 = i1 = i2
-  (* Possibly more efficient version (with a real compiler, at least):
-       if i1.stamp <> 0
-       then i1.stamp = i2.stamp
-       else i2.stamp = 0 && i1.name = i2.name *)
-
-let compare i1 i2 = Stdlib.compare i1 i2
+let same i1 i2 =
+    match i1, i2 with
```
Comment (dra27) [inline typing/ident.ml]:
```suggestion
  match i1, i2 with
```
(whitespace only)

**Review by dra27 — APPROVED** [review]
One misindented line?

**Comment by gasche** [comment]
Can we observe performance benefits coming from this change?

I expect that the new definition of `Ident.same` will be slower in bytecode, and we do run the type-checker compiled in bytecode during a compiler build. I wonder if we can observe a regression there as well, and how clearly it would be offset by reduced native-compiled type-checking times.

**Comment by lpw25** [comment]
A file I happened to be looking at went from ~16s to ~15s.

`compare_val` frequently shows up in the top 2 or 3 in perf profiles and I think this is why.

**Review by xavierleroy — COMMENTED** [review]
The specialization of `compare` to `String.compare` is obviously good.

Concerning `Ident.same`, I am quite certain it will be slower in bytecode and faster in native code.  (I wrote the comment "Possibly more efficient (wilth a real compiler, at least)" many years ago to mean exactly this.)  

If I'm not mistaken, nowadays the `.opt` compilers (compiled to native-code) are installed and used by default whenever available, so for most users of OCaml this should be a gain in compilation times.

Concerning the build of the core system itself, I don't care that much if it gets slightly slower.  With enough cores the build times are negligible anyway.  CI might suffer a bit, though.

In summary: I am in favor provided we can observe a decrease in compilation times with the `.opt` compilers.

**Comment by gasche** [comment]
I don't want to be overly procedural about this but we recently agreed on [Contribution guidelines for optimizations](https://github.com/ocaml/ocaml/blob/trunk/CONTRIBUTING.md#contributing-optimizations). They are more aimed at optimizations to the generated code than to the compiler implementation itself, and here the complexity/bugs downsides are much smaller than in typical optimization passes. But I think that giving some relevant performance numbers to motivate a performance-motivated change would be good hygiene.

**Comment by gasche** [comment]
I looked a bit more at the PR; it was weird that one would need to explicitly turn `compare` into `String.compare` for performance benefits, given that the type-directed-specializer should already do this.

The problem is that those calls to `compare` in the various ident-set-handling functions do not in fact resolve to `Stdlib.compare`, but to an earlier `let compare i1 i2 = Stdlib.compare i1 i2` definition in the module; that one was intended to be at the type `Ident.t -> Ident.t -> int`, but it isn't, and it prevents type-based specialized at the (accidental) callsites. (It was fairly dumb of the person introducing this function to not worry about shadowing below in the module. I can say this frankly as it turns out that it was done by myself in 2015.)

Leo's patch removes this definition of `let compare`... so the further additions of `String.` qualifications to `compare` calls should not be necessary. The reason why removing this `let compare` operation is fine is that Pierre Chambart introduced *another* ident-specialized comparison operation in 2016, also named `compare`, at the *end* of the module (to shadow the previous definition instead of having to deal with shadowing, I guess?).

**Comment by lpw25** [comment]
Interesting point about the performance on bytecode. I was curious so I wrote this micro-benchmark:
```ocaml
open Core_bench

type t =
  | Local of { name: string; stamp: int }
  | Scoped of { name: string; stamp: int; scope: int }
  | Global of string
  | Predef of { name: string; stamp: int }

let same_1 t1 t2 = t1 = t2

let same_2 i1 i2 =
  match i1, i2 with
  | Local { stamp = s1; _ }, Local { stamp = s2; _ }
  | Scoped { stamp = s1; _ }, Scoped { stamp = s2; _ }
  | Predef { stamp = s1; _ }, Predef { stamp = s2; _ } ->
      s1 = s2
  | Global name1, Global name2 ->
      name1 = name2
  | _ ->
      false

let local_1 = Local { name = "local_1"; stamp = 1 }
let local_2 = Local { name = "local_2"; stamp = 2 }
let scoped_1 = Scoped { name = "scoped_1"; stamp = 3; scope = 1 }
let scoped_2 = Scoped { name = "scoped_2"; stamp = 4; scope = 1 }
let predef_1 = Predef { name = "predef_1"; stamp = 5 }
let predef_2 = Predef { name = "predef_2"; stamp = 6 }
let global_1 = Global "global_1"
let global_2 = Global "global_2"

let idents =
  [|local_1;local_2;scoped_1;scoped_2;predef_1;predef_2;global_1;global_2|]

let () = Random.init 1234

let idents =
  Array.init 100 (fun _ ->
    let i = Random.int (Array.length idents - 1) in
    idents.(i))

let main () =
  Core.Command.run (Bench.make_command [
    Bench.Test.create ~name:"same 1" (fun () ->
      for i = 1 to Array.length idents - 1 do
        ignore (same_1 idents.(i) idents.(i-1))
      done);
    Bench.Test.create ~name:"same 2" (fun () ->
      for i = 1 to Array.length idents - 1 do
        ignore (same_2 idents.(i) idents.(i-1))
      done);
  ])

let () = main ()
```
It basically just tries to use the two versions of `same` 100 times on some randomly selected `Ident.t`s.

The results for native-code:
```
┌────────┬────────────┬────────────┐
│ Name   │   Time/Run │ Percentage │
├────────┼────────────┼────────────┤
│ same 1 │ 1_788.51ns │    100.00% │
│ same 2 │   434.79ns │     24.31% │
└────────┴────────────┴────────────┘
```
and for byte-code:
```
┌────────┬──────────┬────────────┐
│ Name   │ Time/Run │ Percentage │
├────────┼──────────┼────────────┤
│ same 1 │   4.10us │    100.00% │
│ same 2 │   3.50us │     85.32% │
└────────┴──────────┴────────────┘
```
So whilst the difference is much smaller on bytecode, the new code is actually still faster.

**Comment by xavierleroy** [comment]
Thanks for the micro-benchmark.  Having done some testing on my own, I confirm that the specialized equality test is no slower than generic equality in bytecode, and clearly faster in native code.

**Review by xavierleroy — APPROVED** [review]
LGTM given the benchmarking results.

**Comment by lpw25** [comment]
Rebased, change entry added, CI happy -- I'm merging.
